Pandas est la librairie Python utilisé pour la manipulation des données tabulaires. Elle est basé sur numpy et propose des méthodes pour importer, analyser, selectionner, filtrer, visualiser et exporter des données.

# Les bases d'un dataframe Python

Le code suivant nous permet d'importer pandas

In [1]:
import pandas as pd

Pour découvrir les capacités de Pandas, nous commencerons pas le **DataFrame** qui est l'objet Pandas qui represente les données tabulaires. Une donnée tabulaire est simplement une donnée sous forme de tableau, qui présente des lignes et des colonnes. Voyons cela avec une donnée réelle.

Nous avons dans le dossier datasets un fichier movie.csv qui contient plusieurs films et leurs caractéristiques.

In [2]:
dossier = 'datasets/'

Pour importer les données dans ce fichier , nous utiliserons la fonction `read_csv` et on lui passe de chemin de fichier `movie.csv`:

In [3]:
dossier + 'movie.csv'

'datasets/movie.csv'

In [4]:
movie = pd.read_csv(dossier + 'movie.csv')

La variable movie contient nos données et elle est de type DataFrame

In [6]:
type(movie)

pandas.core.frame.DataFrame

Affichons cette variable pour voir ce qu'il y a à l'intérieur :

In [7]:
movie

color       director_name  num_critic_for_reviews  duration  \
0               Color       James Cameron                   723.0     178.0   
1               Color      Gore Verbinski                   302.0     169.0   
2               Color          Sam Mendes                   602.0     148.0   
3               Color   Christopher Nolan                   813.0     164.0   
4                 NaN         Doug Walker                     NaN       NaN   
5               Color      Andrew Stanton                   462.0     132.0   
6               Color           Sam Raimi                   392.0     156.0   
7               Color        Nathan Greno                   324.0     100.0   
8               Color         Joss Whedon                   635.0     141.0   
9               Color         David Yates                   375.0     153.0   
10              Color         Zack Snyder                   673.0     183.0   
11              Color        Bryan Singer                   434.0     169.0   
12              Color        Marc Forster                   403.0     106.0   
13              Color      Gore Verbinski                   313.0     151.0   
14              Color      Gore Verbinski                   450.0     150.0   
15              Color         Zack Snyder                   733.0     143.0   
16              Color      Andrew Adamson                   258.0     150.0   
17              Color         Joss Whedon                   703.0     173.0   
18              Color        Rob Marshall                   448.0     136.0   
19              Color    Barry Sonnenfeld                   451.0     106.0   
20              Color       Peter Jackson                   422.0     164.0   
21              Color           Marc Webb                   599.0     153.0   
22              Color        Ridley Scott                   343.0     156.0   
23              Color       Peter Jackson                   509.0     186.0   
24              Color         Chris Weitz                   251.0     113.0   
25              Color       Peter Jackson                   446.0     201.0   
26              Color       James Cameron                   315.0     194.0   
27              Color       Anthony Russo                   516.0     147.0   
28              Color          Peter Berg                   377.0     131.0   
29              Color     Colin Trevorrow                   644.0     124.0   
...               ...                 ...                     ...       ...   
4886            Color          Eric Eason                    28.0      79.0   
4887            Color            Uwe Boll                    58.0      80.0   
4888  Black and White   Richard Linklater                    61.0     100.0   
4889            Color     Joseph Mazzella                     NaN      90.0   
4890            Color        Travis Legge                     1.0      90.0   
4891            Color       Alex Kendrick                     5.0     120.0   
4892            Color       Marcus Nispel                    43.0      91.0   
4893              NaN     Brandon Landers                     NaN     143.0   
4894            Color         Jay Duplass                    51.0      85.0   
4895  Black and White          Jim Chuchu                     6.0      60.0   
4896            Color          Daryl Wein                    22.0      88.0   
4897            Color         Jason Trost                    42.0      78.0   
4898            Color         John Waters                    73.0     108.0   
4899            Color     Olivier Assayas                    81.0     110.0   
4900            Color        Jafar Panahi                    64.0      90.0   
4901  Black and White       Ivan Kavanagh                    12.0      83.0   
4902            Color    Kiyoshi Kurosawa                    78.0     111.0   
4903            Color        Tadeo Garcia                     NaN      84.0   
4904            Color  Thomas L. Phillips                    13.0 

On peut remarquer que les colonnes sont identifiables par leur nom sur la première ligne et que les lignes sont identifiables par un numéro unique à gauche. La colonne qui contient les numeros des lignes est appélé index du DataFrame.

Nous pouvons jeter un coup d'oeil aux 5 premières lignes de notre Dataframe avec la fonction `head`:

In [8]:
movie.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

Affichons la liste des colonnes : 

In [9]:
movie.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [10]:
type(movie.columns)

pandas.core.indexes.base.Index

Affichons l'index : 

In [11]:
movie.index

RangeIndex(start=0, stop=4916, step=1)

Nous pouvons avoir accès aux valeurs du DataFrame Sous forme de tableau numpy avec l'attribut **values** : 

In [12]:
movie.values

array([['Color', 'James Cameron', 723.0, ..., 7.9, 1.78, 33000],
       ['Color', 'Gore Verbinski', 302.0, ..., 7.1, 2.35, 0],
       ['Color', 'Sam Mendes', 602.0, ..., 6.8, 2.35, 85000],
       ...,
       ['Color', 'Benjamin Roberds', 13.0, ..., 6.3, nan, 16],
       ['Color', 'Daniel Hsia', 14.0, ..., 6.3, 2.35, 660],
       ['Color', 'Jon Gunn', 43.0, ..., 6.6, 1.85, 456]], dtype=object)

In [13]:
movie.values.shape

(4916, 28)

In [14]:
movie.values.ndim

2

Vous avez dû remarquer dans le Dataframe (qu'on va désormais appeler df), que certaines cellules contiennent **NaN**. Cela veut dire **Not a Number**. C'est la manière de Pandas de signifier que la valeur est manquante.

# Quelques fonctions et méthodes pour mieux comprendre la data

Nous avons déjà vu la fonction head qui permet d'afficher les 5 premières lignes du df. L'inverse est possible avec la fonction tail pour afficher les 5 dernières lignes de notre dataframe.

In [15]:
movie.tail()

color     director_name  num_critic_for_reviews  duration  \
4911  Color       Scott Smith                     1.0      87.0   
4912  Color               NaN                    43.0      43.0   
4913  Color  Benjamin Roberds                    13.0      76.0   
4914  Color       Daniel Hsia                    14.0     100.0   
4915  Color          Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
4911                      2.0                   318.0     Daphne Zuniga   
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes    gross                        genres  ...  \
4911                   637.0      NaN                  Comedy|Drama  ...   
4912                   841.0      NaN  Crime|Drama|Mystery|Thriller  ...   
4913                     0.0      NaN         Drama|Horror|Thriller  ...   
4914                   946.0  10443.0          Comedy|Drama|Romance  ...   
4915                    86.0  85222.0                   Documentary  ...   

     num_user_for_reviews language  country  content_rating  budget  \
4911                  6.0  English   Canada             NaN     NaN   
4912                359.0  English      USA           TV-14     NaN   
4913                  3.0  English      USA             NaN  1400.0   
4914                  9.0  English      USA           PG-13     NaN   
4915                 84.0  English      USA              PG  1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
4911      2013.0                  470.0        7.7           NaN   
4912         NaN                  593.0        7.5         16.00   
4913      2013.0                    0.0        6.3           NaN   
4914      2012.0                  719.0        6.3          2.35   
4915      2004.0                   23.0        6.6          1.85   

     movie_facebook_likes  
4911                   84  
4912                32000  
4913                   16  
4914                  660  
4915                  456  

[5 rows x 28 columns]

Nous pouvons aussi quelques lignes prises aléatoirement : 

In [17]:
movie.sample(5)

color       director_name  num_critic_for_reviews  duration  \
1477  Color           Hark Tsui                    67.0      91.0   
3079  Color  Olatunde Osunsanmi                   183.0      98.0   
3714  Color          Billy Kent                    50.0      88.0   
4624  Color       Joseph Dorman                    29.0      93.0   
391   Color       George Miller                   206.0     108.0   

      director_facebook_likes  actor_3_facebook_likes   actor_2_name  \
1477                     85.0                   105.0    Lela Rochon   
3079                     58.0                   242.0    Will Patton   
3714                      0.0                   276.0       Tim Russ   
4624                      0.0                    59.0  Peter Riegert   
391                     750.0                   971.0   Hugh Jackman   

      actor_1_facebook_likes        gross  \
1477                   635.0   10076136.0   
3079                 14000.0   25464480.0   
3714                   740.0          NaN   
4624                   399.0     906666.0   
391                  49000.0  197992827.0   

                                     genres  ... num_user_for_reviews  \
1477                 Action|Comedy|Thriller  ...                141.0   
3079                Mystery|Sci-Fi|Thriller  ...                330.0   
3714                   Comedy|Drama|Romance  ...                 56.0   
4624                            Documentary  ...                  5.0   
391   Animation|Comedy|Family|Music|Romance  ...                548.0   

     language  country  content_rating       budget  title_year  \
1477  English    Aruba               R   35000000.0      1998.0   
3079  English      USA           PG-13   10000000.0      2009.0   
3714  English      USA               R    5000000.0      2006.0   
4624  English      USA         Unrated     500000.0      2011.0   
391   English      USA              PG  100000000.0      2006.0   

     actor_2_facebook_likes imdb_score  aspect_ratio movie_facebook_likes  
1477                  316.0        4.8          2.35                  471  
3079                  537.0        5.9          2.35                15000  
3714                  625.0        5.6          2.35                  255  
4624                  169.0        6.8         16.00                  287  
391                 20000.0        6.5          2.35                    0  

[5 rows x 28 columns]

Nous pouvons avoir accès au nombre de lignes et de colonnes du df  :

In [19]:
movie.shape

(4916, 28)

Nous avons 4916 et 28 colonnes. Jetons un coup d'oeil au type de données dans chaque variable : 

In [21]:
movie.dtypes

color                         object
director_name                 object
num_critic_for_reviews       float64
duration                     float64
director_facebook_likes      float64
actor_3_facebook_likes       float64
actor_2_name                  object
actor_1_facebook_likes       float64
gross                        float64
genres                        object
actor_1_name                  object
movie_title                   object
num_voted_users                int64
cast_total_facebook_likes      int64
actor_3_name                  object
facenumber_in_poster         float64
plot_keywords                 object
movie_imdb_link               object
num_user_for_reviews         float64
language                      object
country                       object
content_rating                object
budget                       float64
title_year                   float64
actor_2_facebook_likes       float64
imdb_score                   float64
aspect_ratio                 float64
m

Le type object est le type utilisé par Python quand une colonne contient des chaines de caractères ou des données de plusieurs types ou simplement lorsque des valeurs sont manquantes.

Nous pouvons afficher le nombre de valeurs non manquantes dans chaque colonne avec leur type : 

In [22]:
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 28 columns):
color                        4897 non-null object
director_name                4814 non-null object
num_critic_for_reviews       4867 non-null float64
duration                     4901 non-null float64
director_facebook_likes      4814 non-null float64
actor_3_facebook_likes       4893 non-null float64
actor_2_name                 4903 non-null object
actor_1_facebook_likes       4909 non-null float64
gross                        4054 non-null float64
genres                       4916 non-null object
actor_1_name                 4909 non-null object
movie_title                  4916 non-null object
num_voted_users              4916 non-null int64
cast_total_facebook_likes    4916 non-null int64
actor_3_name                 4893 non-null object
facenumber_in_poster         4903 non-null float64
plot_keywords                4764 non-null object
movie_imdb_link              4916 non-

In [18]:
movie.color

0                 Color
1                 Color
2                 Color
3                 Color
4                   NaN
5                 Color
6                 Color
7                 Color
8                 Color
9                 Color
10                Color
11                Color
12                Color
13                Color
14                Color
15                Color
16                Color
17                Color
18                Color
19                Color
20                Color
21                Color
22                Color
23                Color
24                Color
25                Color
26                Color
27                Color
28                Color
29                Color
             ...       
4886              Color
4887              Color
4888    Black and White
4889              Color
4890              Color
4891              Color
4892              Color
4893                NaN
4894              Color
4895    Black and White
4896            

In [19]:
movie.color.value_counts()

Color              4693
Black and White     204
Name: color, dtype: int64

In [20]:
movie.rename(columns = {'director_facebook_likes':'director_fb_likes'}, inplace=True)

In [21]:
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 28 columns):
color                        4897 non-null object
director_name                4814 non-null object
num_critic_for_reviews       4867 non-null float64
duration                     4901 non-null float64
director_fb_likes            4814 non-null float64
actor_3_facebook_likes       4893 non-null float64
actor_2_name                 4903 non-null object
actor_1_facebook_likes       4909 non-null float64
gross                        4054 non-null float64
genres                       4916 non-null object
actor_1_name                 4909 non-null object
movie_title                  4916 non-null object
num_voted_users              4916 non-null int64
cast_total_facebook_likes    4916 non-null int64
actor_3_name                 4893 non-null object
facenumber_in_poster         4903 non-null float64
plot_keywords                4764 non-null object
movie_imdb_link              4916 non-

In [22]:
movie_string = movie.select_dtypes('object')

In [23]:
movie_string.head()

color      director_name      actor_2_name  \
0  Color      James Cameron  Joel David Moore   
1  Color     Gore Verbinski     Orlando Bloom   
2  Color         Sam Mendes      Rory Kinnear   
3  Color  Christopher Nolan    Christian Bale   
4    NaN        Doug Walker        Rob Walker   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
4                      Documentary      Doug Walker   

                                  movie_title          actor_3_name  \
0                                      Avatar             Wes Studi   
1    Pirates of the Caribbean: At World's End        Jack Davenport   
2                                     Spectre      Stephanie Sigman   
3                       The Dark Knight Rises  Joseph Gordon-Levitt   
4  Star Wars: Episode VII - The Force Awakens                   NaN   

                                       plot_keywords  \
0             avatar|future|marine|native|paraplegic   
1  goddess|marriage ceremony|marriage proposal|pi...   
2                bomb|espionage|sequel|spy|terrorist   
3  deception|imprisonment|lawlessness|police offi...   
4                                                NaN   

                                     movie_imdb_link language country  \
0  http://www.imdb.com/title/tt0499549/?ref_=fn_t...  English     USA   
1  http://www.imdb.com/title/tt0449088/?ref_=fn_t...  English     USA   
2  http://www.imdb.com/title/tt2379713/?ref_=fn_t...  English      UK   
3  http://www.imdb.com/title/tt1345836/?ref_=fn_t...  English     USA   
4  http://www.imdb.com/title/tt5289954/?ref_=fn_t...      NaN     NaN   

  content_rating  
0          PG-13  
1          PG-13  
2          PG-13  
3          PG-13  
4            NaN

In [24]:
movie_number = movie.select_dtypes(include='number')

In [25]:
movie_number.head()

num_critic_for_reviews  duration  director_fb_likes  \
0                   723.0     178.0                0.0   
1                   302.0     169.0              563.0   
2                   602.0     148.0                0.0   
3                   813.0     164.0            22000.0   
4                     NaN       NaN              131.0   

   actor_3_facebook_likes  actor_1_facebook_likes        gross  \
0                   855.0                  1000.0  760505847.0   
1                  1000.0                 40000.0  309404152.0   
2                   161.0                 11000.0  200074175.0   
3                 23000.0                 27000.0  448130642.0   
4                     NaN                   131.0          NaN   

   num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
0           886204                       4834                   0.0   
1           471220                      48350                   0.0   
2           275868                      11700                   1.0   
3          1144337                     106759                   0.0   
4                8                        143                   0.0   

   num_user_for_reviews       budget  title_year  actor_2_facebook_likes  \
0                3054.0  237000000.0      2009.0                   936.0   
1                1238.0  300000000.0      2007.0                  5000.0   
2                 994.0  245000000.0      2015.0                   393.0   
3                2701.0  250000000.0      2012.0                 23000.0   
4                   NaN          NaN         NaN                    12.0   

   imdb_score  aspect_ratio  movie_facebook_likes  
0         7.9          1.78                 33000  
1         7.1          2.35                     0  
2         6.8          2.35                 85000  
3         8.5          2.35                164000  
4         7.1           NaN                     0

In [26]:
movie_number.describe()

num_critic_for_reviews     duration  director_fb_likes  \
count             4867.000000  4901.000000        4814.000000   
mean               137.988905   107.090798         691.014541   
std                120.239379    25.286015        2832.954125   
min                  1.000000     7.000000           0.000000   
25%                 49.000000    93.000000           7.000000   
50%                108.000000   103.000000          48.000000   
75%                191.000000   118.000000         189.750000   
max                813.000000   511.000000       23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             4893.000000             4909.000000  4.054000e+03   
mean               631.276313             6494.488491  4.764451e+07   
std               1625.874802            15106.986884  6.737255e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                132.000000              607.000000  5.019656e+06   
50%                366.000000              982.000000  2.504396e+07   
75%                633.000000            11000.000000  6.110841e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     4.916000e+03                4916.000000           4903.000000   
mean      8.264492e+04                9579.815907              1.377320   
std       1.383222e+05               18164.316990              2.023826   
min       5.000000e+00                   0.000000              0.000000   
25%       8.361750e+03                1394.750000              0.000000   
50%       3.313250e+04                3049.000000              1.000000   
75%       9.377275e+04               13616.750000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           4895.000000  4.432000e+03  4810.000000   
mean             267.668846  3.654749e+07  2002.447609   
std              372.934839  1.002427e+08    12.453977   
min                1.000000  2.180000e+02  1916.000000   
25%               64.000000  6.000000e+06  1999.000000   
50%              153.000000  1.985000e+07  2005.000000   
75%              320.500000  4.300000e+07  2011.000000   
max             5060.000000  4.200000e+09  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             4903.000000  4916.000000   4590.000000           4916.000000  
mean              1621.923516     6.437429      2.222349           7348.294142  
std               4011.299523     1.127802      1.402940          19206.016458  
min                  0.000000     1.600000      1.180000              0.000000  
25%                277.000000     5.800000      1.850000              0.000000  
50%                593.000000     6.600000      2.350000            159.000000  
75%                912.000000     7.200000      2.350000           2000.000000  
max             137000.000000     9.500000     16.000000         349000.000000

In [27]:

cols_interest = ["director_name", "duration", "gross", "genres",
                 'movie_title', "plot_keywords","budget", "title_year", "language", "imdb_score"]
movie_reduced = movie[cols_interest]

In [28]:
movie_reduced.sample(2)

director_name  duration       gross                 genres  \
755   Jean-Pierre Jeunet     116.0  47748610.0   Action|Horror|Sci-Fi   
1550      Norman Jewison     110.0   4357000.0  Comedy|Family|Fantasy   

              movie_title                                      plot_keywords  \
755   Alien: Resurrection               alien|blood|breeding|clone|scientist   
1550                Bogus  box office flop|circus|imaginary friend|magici...   

          budget  title_year language  imdb_score  
755   75000000.0      1997.0  English         6.3  
1550  30000000.0      1996.0  English         5.3

In [29]:
movie_reduced.drop(0, inplace=True)

/home/kevin/miniconda3/envs/ds/lib/python3.7/site-packages/pandas/core/frame.py:3940: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [30]:
movie_reduced = movie_reduced.drop(['movie_title', 'plot_keywords'], axis = 1)

In [31]:
movie_reduced.head()

director_name  duration        gross                     genres  \
1     Gore Verbinski     169.0  309404152.0   Action|Adventure|Fantasy   
2         Sam Mendes     148.0  200074175.0  Action|Adventure|Thriller   
3  Christopher Nolan     164.0  448130642.0            Action|Thriller   
4        Doug Walker       NaN          NaN                Documentary   
5     Andrew Stanton     132.0   73058679.0    Action|Adventure|Sci-Fi   

        budget  title_year language  imdb_score  
1  300000000.0      2007.0  English         7.1  
2  245000000.0      2015.0  English         6.8  
3  250000000.0      2012.0  English         8.5  
4          NaN         NaN      NaN         7.1  
5  263700000.0      2012.0  English         6.6

In [32]:
movie_reduced['language'].value_counts()

English       4581
French          73
Spanish         40
Hindi           28
Mandarin        24
German          19
Japanese        17
Cantonese       11
Italian         11
Russian         11
Portuguese       8
Korean           7
Hebrew           5
Danish           5
Swedish          5
Arabic           5
Dutch            4
Persian          4
Norwegian        4
Polish           3
Thai             3
Chinese          3
None             2
Romanian         2
Indonesian       2
Zulu             2
Dari             2
Icelandic        2
Aboriginal       2
Panjabi          1
Mongolian        1
Bosnian          1
Swahili          1
Hungarian        1
Telugu           1
Czech            1
Vietnamese       1
Filipino         1
Kazakh           1
Aramaic          1
Tamil            1
Kannada          1
Urdu             1
Slovenian        1
Maya             1
Greek            1
Dzongkha         1
Name: language, dtype: int64

In [33]:
movie_reduced['language'].value_counts(ascending=True)

Dzongkha         1
Greek            1
Maya             1
Slovenian        1
Urdu             1
Kannada          1
Tamil            1
Aramaic          1
Kazakh           1
Filipino         1
Vietnamese       1
Czech            1
Telugu           1
Hungarian        1
Swahili          1
Bosnian          1
Mongolian        1
Panjabi          1
Aboriginal       2
Icelandic        2
Dari             2
Zulu             2
Indonesian       2
Romanian         2
None             2
Chinese          3
Thai             3
Polish           3
Norwegian        4
Persian          4
Dutch            4
Arabic           5
Swedish          5
Danish           5
Hebrew           5
Korean           7
Portuguese       8
Russian         11
Italian         11
Cantonese       11
Japanese        17
German          19
Mandarin        24
Hindi           28
Spanish         40
French          73
English       4581
Name: language, dtype: int64

In [34]:
(movie_reduced['language'].value_counts() / len(movie_reduced)) * 100

English       93.204476
French         1.485249
Spanish        0.813835
Hindi          0.569685
Mandarin       0.488301
German         0.386572
Japanese       0.345880
Cantonese      0.223805
Italian        0.223805
Russian        0.223805
Portuguese     0.162767
Korean         0.142421
Hebrew         0.101729
Danish         0.101729
Swedish        0.101729
Arabic         0.101729
Dutch          0.081384
Persian        0.081384
Norwegian      0.081384
Polish         0.061038
Thai           0.061038
Chinese        0.061038
None           0.040692
Romanian       0.040692
Indonesian     0.040692
Zulu           0.040692
Dari           0.040692
Icelandic      0.040692
Aboriginal     0.040692
Panjabi        0.020346
Mongolian      0.020346
Bosnian        0.020346
Swahili        0.020346
Hungarian      0.020346
Telugu         0.020346
Czech          0.020346
Vietnamese     0.020346
Filipino       0.020346
Kazakh         0.020346
Aramaic        0.020346
Tamil          0.020346
Kannada        0

In [35]:
del movie_reduced['genres']

In [36]:
movie_reduced['title'] = movie['movie_title']

In [37]:
movie_reduced.sample(2)

director_name  duration       gross       budget  title_year language  \
3476  Kenny Ortega     111.0         NaN    7000000.0      2007.0  English   
246     Josh Trank     100.0  56114221.0  120000000.0      2015.0  English   

      imdb_score                  title  
3476         4.8  High School Musical 2  
246          4.3         Fantastic Four

In [38]:
movie_reduced.rename(columns={'gross': 'revenue'}, inplace=True)

In [39]:
movie_reduced.sample(2)

director_name  duration     revenue      budget  title_year language  \
1455  John Singleton     106.0  28064226.0  35000000.0      2011.0  English   
862   Akiva Schaffer     102.0  34350553.0  68000000.0      2012.0  English   

      imdb_score      title  
1455         5.0  Abduction  
862          5.7  The Watch

In [40]:
movie_reduced.columns.tolist()

['director_name',
 'duration',
 'revenue',
 'budget',
 'title_year',
 'language',
 'imdb_score',
 'title']

In [41]:
new_arrangement = sorted(movie_reduced.columns.tolist())
movie_reduced = movie_reduced[new_arrangement]

In [42]:
movie_reduced.head(2)

budget   director_name  duration  imdb_score language      revenue  \
1  300000000.0  Gore Verbinski     169.0         7.1  English  309404152.0   
2  245000000.0      Sam Mendes     148.0         6.8  English  200074175.0   

                                      title  title_year  
1  Pirates of the Caribbean: At World's End      2007.0  
2                                   Spectre      2015.0

In [43]:
movie_reduced['revenue'].max()

658672302.0

In [46]:
np.argmax(movie_reduced.revenue)

/home/kevin/miniconda3/envs/ds/lib/python3.7/site-packages/numpy/core/fromnumeric.py:61: FutureWarning: 
The current behaviour of 'Series.argmax' is deprecated, use 'idxmax'
instead.
The behavior of 'argmax' will be corrected to return the positional
maximum in the future. For now, use 'series.values.argmax' or
'np.argmax(np.array(values))' to get the position of the maximum
row.
  return bound(*args, **kwds)


26

In [49]:
movie_reduced.loc[26]

budget                   2e+08
director_name    James Cameron
duration                   194
imdb_score                 7.7
language               English
revenue            6.58672e+08
title                  Titanic
title_year                1997
Name: 26, dtype: object

In [50]:
movie_reduced.sort_values('revenue', ascending=False)[:6]

budget      director_name  duration  imdb_score language  \
26    200000000.0      James Cameron     194.0         7.7  English   
29    150000000.0    Colin Trevorrow     124.0         7.0  English   
17    220000000.0        Joss Whedon     173.0         8.1  English   
66    185000000.0  Christopher Nolan     152.0         9.0  English   
237   115000000.0       George Lucas     136.0         6.5  English   
2973   11000000.0       George Lucas     125.0         8.7  English   

          revenue                                      title  title_year  
26    658672302.0                                    Titanic      1997.0  
29    652177271.0                             Jurassic World      2015.0  
17    623279547.0                               The Avengers      2012.0  
66    533316061.0                            The Dark Knight      2008.0  
237   474544677.0  Star Wars: Episode I - The Phantom Menace      1999.0  
2973  460935665.0         Star Wars: Episode IV - A New Hope      1977.0

In [51]:
movie_reduced.sort_values(['title_year', 'imdb_score'], ascending=False)

budget        director_name  duration  imdb_score  language  \
4312   17000000.0       John Stockwell      90.0         9.1       NaN   
4277          NaN     Mitchell Altieri      87.0         8.7   English   
3798    4400000.0           Raja Menon     130.0         8.5     Hindi   
27    250000000.0        Anthony Russo     147.0         8.2   English   
98            NaN         Hideaki Anno     120.0         8.2  Japanese   
803    58000000.0           Tim Miller     108.0         8.1   English   
3367    8000000.0      Bruce Beresford     104.0         8.0   English   
79    175000000.0          Jon Favreau     106.0         7.8   English   
1186   40000000.0            James Wan     134.0         7.8   English   
2355          NaN        James Schamus     110.0         7.8    Hebrew   
3312          NaN       Justin Tipping      80.0         7.8   English   
3252          NaN          Hong-jin Na     156.0         7.7    Korean   
2152   20000000.0        Thea Sharrock     110.0         7.6   English   
57    185000000.0           Justin Lin     122.0         7.5   English   
1717   19000000.0         Greg Tiernan      89.0         7.5   English   
1927   23000000.0      Dexter Fletcher     106.0         7.5   English   
961    50000000.0          Michael Bay     144.0         7.4   English   
4659     500000.0  Darren Lynn Bousman      97.0         7.4   English   
65    178000000.0         Bryan Singer     144.0         7.3   English   
108   160000000.0         Duncan Jones     123.0         7.3   English   
1062   25000000.0          Brad Furman     127.0         7.3   English   
3075   65000000.0         David Lowery     102.0         7.3   English   
3629   15000000.0     Dan Trachtenberg     104.0         7.3   English   
153   145000000.0   Alessandro Carloni      95.0         7.2   English   
2322   20000000.0  Jonathan Jakubowicz     105.0         7.2   English   
3477    8700000.0           Kim Nguyen      96.0         7.2   English   
4605          NaN         Thomas Lilti     102.0         7.2    French   
222   120000000.0      Paul Greengrass     123.0         7.1   English   
1736   29000000.0       Stephen Frears     110.0         7.1   English   
2211   20000000.0          Henry Joost      96.0         7.1   English   
...           ...                  ...       ...         ...       ...   
4196          NaN                  NaN      60.0         7.3   English   
4234          NaN                  NaN      30.0         7.3   English   
4362          NaN                  NaN     120.0         7.3   English   
4383          NaN                  NaN      27.0         7.3   English   
276           NaN  Christopher Barnard      22.0         7.2       NaN   
1955          NaN                  NaN      30.0         7.2   English   
2313          NaN                  NaN      30.0         7.2   English   
4             NaN          Doug Walker       NaN         7.1       NaN   
3315          NaN                  NaN      60.0         7.1   English   
4416          NaN                  NaN       NaN         7.1   English   
817           NaN                  NaN      30.0         7.0   English   
1014          NaN                  NaN      43.0         7.0   English   
1431          NaN                  NaN      60.0         7.0   English   
2214          NaN                  NaN      24.0         7.0  Japanese   
2053          NaN                  NaN      50.0         6.8   English   
824           NaN                  NaN      22.0         6.7   English   
2230          NaN                  NaN      30.0         6.7   English   
2429          NaN                  NaN      60.0         6.7       NaN   
4704          NaN                  NaN      43.0         6.7   English   
807     3000000.0                  NaN      22.0         6.6   English   
2341          NaN                  NaN      43.0         6.6   English   
2518          NaN                  NaN      60.0         6.6   English   
2733          N

Nous pouvons déjà remarquer que certaines colonnes ont plusieurs valeurs manquantes. Nous y reviendrons plus bas.

# Selectionner une colonne d'un df

Nous souhaitons voir les données d'une colonne particulière en la sélectionnant. Considérons la colonne movie_title qui contient le titre des films dans notre DataFrame:

In [23]:
movie.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [24]:
movie['movie_title']

0                                            Avatar
1          Pirates of the Caribbean: At World's End
2                                           Spectre
3                             The Dark Knight Rises
4        Star Wars: Episode VII - The Force Awakens
5                                       John Carter
6                                      Spider-Man 3
7                                           Tangled
8                           Avengers: Age of Ultron
9            Harry Potter and the Half-Blood Prince
10               Batman v Superman: Dawn of Justice
11                                 Superman Returns
12                                Quantum of Solace
13       Pirates of the Caribbean: Dead Man's Chest
14                                  The Lone Ranger
15                                     Man of Steel
16         The Chronicles of Narnia: Prince Caspian
17                                     The Avengers
18      Pirates of the Caribbean: On Stranger Tides
19          

Une colonne est appelée Série dans Pandas. Plusieurs fonctions de Pandas  applicables aux DataFrames le sont aussi pour les Séries mais pas toutes.

In [26]:
type(movie['movie_title'])

pandas.core.series.Series

Pour sélectionner une colonne, il faut donc mettre son nom entre `[ ]` avec le nom du df à gauche. Cela vous rapelle probablement la sélection de valeur dans un dictionnaire Python:

In [29]:
un_dict = {"titre": "Seigneur des anneaux"}

In [30]:
un_dict['titre']

'Seigneur des anneaux'

# Sélectionner plusieurs colonnes d'un DataFrame

Ici, nous souhaitons sélectionner les colonne `movie_title` ,  `language` et `director_name` pour répondre à une question d'analyse ou pour une visualition:

In [31]:
movie.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [33]:
liste_des_colonnes = ['movie_title', 'language', 'director_name']
movie[liste_des_colonnes]

movie_title  language  \
0                                          Avatar   English   
1        Pirates of the Caribbean: At World's End   English   
2                                         Spectre   English   
3                           The Dark Knight Rises   English   
4      Star Wars: Episode VII - The Force Awakens       NaN   
5                                     John Carter   English   
6                                    Spider-Man 3   English   
7                                         Tangled   English   
8                         Avengers: Age of Ultron   English   
9          Harry Potter and the Half-Blood Prince   English   
10             Batman v Superman: Dawn of Justice   English   
11                               Superman Returns   English   
12                              Quantum of Solace   English   
13     Pirates of the Caribbean: Dead Man's Chest   English   
14                                The Lone Ranger   English   
15                                   Man of Steel   English   
16       The Chronicles of Narnia: Prince Caspian   English   
17                                   The Avengers   English   
18    Pirates of the Caribbean: On Stranger Tides   English   
19                                 Men in Black 3   English   
20      The Hobbit: The Battle of the Five Armies   English   
21                         The Amazing Spider-Man   English   
22                                     Robin Hood   English   
23            The Hobbit: The Desolation of Smaug   English   
24                             The Golden Compass   English   
25                                      King Kong   English   
26                                        Titanic   English   
27                     Captain America: Civil War   English   
28                                     Battleship   English   
29                                 Jurassic World   English   
...                                           ...       ...   
4886                                       Manito   English   
4887                                      Rampage   English   
4888                                      Slacker   English   
4889                                  Dutch Kills   English   
4890                                    Dry Spell   English   
4891                                     Flywheel   English   
4892                                       Exeter   English   
4893                                   The Ridges   English   
4894                              The Puffy Chair   English   
4895                         Stories of Our Lives   Swahili   
4896                             Breaking Upwards   English   
4897                     All Superheroes Must Die   English   
4898                               Pink Flamingos   English   
4899                                        Clean    French   
4900                                   The Circle   Persian   
4901                                  Tin Can Man   English   
4902                                     The Cure  Japanese   
4903                               On the Downlow   English   
4904                 Sanctuary; Quite a Conundrum   English   
4905                                         Bang   English   
4906                                       Primer   English   
4907                                       Cavite   English   
4908                                  El Mariachi   Spanish   
4909                              The Mongol King   English   
4910                                    Newlyweds   English   
4911                      Signed Sealed Delivered   English   
4912                                The Following   English   
4913                         A Plague So Pleasant   English   
4914                             Shanghai Calling   English   
4915                            My Date with Drew   English   

           director_name  
0          James Cameron  
1         Gore Verbinski  
2             Sam Mendes  
3      Christopher Nola

Pour le faire nous avons créer une liste des noms des colonnes puis avons fait la sélection en mettrant la liste en les crochets `[ ]`. Nous pouvons le faire directement sans passer par une liste aléatoire : 

In [34]:
movie[ ['movie_title', 'language', 'director_name'] ]

movie_title  language  \
0                                          Avatar   English   
1        Pirates of the Caribbean: At World's End   English   
2                                         Spectre   English   
3                           The Dark Knight Rises   English   
4      Star Wars: Episode VII - The Force Awakens       NaN   
5                                     John Carter   English   
6                                    Spider-Man 3   English   
7                                         Tangled   English   
8                         Avengers: Age of Ultron   English   
9          Harry Potter and the Half-Blood Prince   English   
10             Batman v Superman: Dawn of Justice   English   
11                               Superman Returns   English   
12                              Quantum of Solace   English   
13     Pirates of the Caribbean: Dead Man's Chest   English   
14                                The Lone Ranger   English   
15                                   Man of Steel   English   
16       The Chronicles of Narnia: Prince Caspian   English   
17                                   The Avengers   English   
18    Pirates of the Caribbean: On Stranger Tides   English   
19                                 Men in Black 3   English   
20      The Hobbit: The Battle of the Five Armies   English   
21                         The Amazing Spider-Man   English   
22                                     Robin Hood   English   
23            The Hobbit: The Desolation of Smaug   English   
24                             The Golden Compass   English   
25                                      King Kong   English   
26                                        Titanic   English   
27                     Captain America: Civil War   English   
28                                     Battleship   English   
29                                 Jurassic World   English   
...                                           ...       ...   
4886                                       Manito   English   
4887                                      Rampage   English   
4888                                      Slacker   English   
4889                                  Dutch Kills   English   
4890                                    Dry Spell   English   
4891                                     Flywheel   English   
4892                                       Exeter   English   
4893                                   The Ridges   English   
4894                              The Puffy Chair   English   
4895                         Stories of Our Lives   Swahili   
4896                             Breaking Upwards   English   
4897                     All Superheroes Must Die   English   
4898                               Pink Flamingos   English   
4899                                        Clean    French   
4900                                   The Circle   Persian   
4901                                  Tin Can Man   English   
4902                                     The Cure  Japanese   
4903                               On the Downlow   English   
4904                 Sanctuary; Quite a Conundrum   English   
4905                                         Bang   English   
4906                                       Primer   English   
4907                                       Cavite   English   
4908                                  El Mariachi   Spanish   
4909                              The Mongol King   English   
4910                                    Newlyweds   English   
4911                      Signed Sealed Delivered   English   
4912                                The Following   English   
4913                         A Plague So Pleasant   English   
4914                             Shanghai Calling   English   
4915                            My Date with Drew   English   

           director_name  
0          James Cameron  
1         Gore Verbinski  
2             Sam Mendes  
3      Christopher Nola

Le résultat est un dataframe avec 3 colonnes:

In [35]:
type(movie[ ['movie_title', 'language', 'director_name'] ])

pandas.core.frame.DataFrame

# Selectioner une ligne d'un dataframe avec iloc

Nous souhaitons sélectionner une ligne particulière de notre dataframe. Les lignes sont identifiés grâce à leur index. Il nous suffit de connaitre l'index qui correspond à notre ligne.

Sélectionnons la 5ème ligne du dataframe. Puisque les index commencent par 0, la 5ème ligne a l'index 4:

In [36]:
movie.iloc[4]

color                                                                      NaN
director_name                                                      Doug Walker
num_critic_for_reviews                                                     NaN
duration                                                                   NaN
director_facebook_likes                                                    131
actor_3_facebook_likes                                                     NaN
actor_2_name                                                        Rob Walker
actor_1_facebook_likes                                                     131
gross                                                                      NaN
genres                                                             Documentary
actor_1_name                                                       Doug Walker
movie_title                         Star Wars: Episode VII - The Force Awakens
num_voted_users                                     

Nous avons à gauche le nom des colonnes et à droite la valeur correspondante pour la 5ème ligne. Le tout est une Série.

In [37]:
type(movie.iloc[4])

pandas.core.series.Series

# Sélectionner plusieurs lignes d'un DataFrame

Si vous avez suivi la partie Numpy de ce bootcamp, cette partie sera facile à comprendre.

Nous souhaitons sélectionner les lignes dont les index sont 4, 5, 6, 7.

In [38]:
movie.iloc[4:8]

color   director_name  num_critic_for_reviews  duration  \
4    NaN     Doug Walker                     NaN       NaN   
5  Color  Andrew Stanton                   462.0     132.0   
6  Color       Sam Raimi                   392.0     156.0   
7  Color    Nathan Greno                   324.0     100.0   

   director_facebook_likes  actor_3_facebook_likes     actor_2_name  \
4                    131.0                     NaN       Rob Walker   
5                    475.0                   530.0  Samantha Morton   
6                      0.0                  4000.0     James Franco   
7                     15.0                   284.0     Donna Murphy   

   actor_1_facebook_likes        gross  \
4                   131.0          NaN   
5                   640.0   73058679.0   
6                 24000.0  336530303.0   
7                   799.0  200807262.0   

                                              genres  ...  \
4                                        Documentary  ...   
5                            Action|Adventure|Sci-Fi  ...   
6                           Action|Adventure|Romance  ...   
7  Adventure|Animation|Comedy|Family|Fantasy|Musi...  ...   

  num_user_for_reviews language  country  content_rating       budget  \
4                  NaN      NaN      NaN             NaN          NaN   
5                738.0  English      USA           PG-13  263700000.0   
6               1902.0  English      USA           PG-13  258000000.0   
7                387.0  English      USA              PG  260000000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
4         NaN                   12.0        7.1           NaN   
5      2012.0                  632.0        6.6          2.35   
6      2007.0                11000.0        6.2          2.35   
7      2010.0                  553.0        7.8          1.85   

  movie_facebook_likes  
4                    0  
5                24000  
6                    0  
7                29000  

[4 rows x 28 columns]

Séléctionnons les lignes dont les index sont 4, 45, 89, 12: 

In [39]:
movie.iloc[ [4, 45, 89, 12] ]

color director_name  num_critic_for_reviews  duration  \
4     NaN   Doug Walker                     NaN       NaN   
45  Color     James Wan                   424.0     140.0   
89  Color      Don Hall                   384.0     102.0   
12  Color  Marc Forster                   403.0     106.0   

    director_facebook_likes  actor_3_facebook_likes     actor_2_name  \
4                     131.0                     NaN       Rob Walker   
45                      0.0                 14000.0      Paul Walker   
89                     38.0                   562.0    Daniel Henney   
12                    395.0                   393.0  Mathieu Amalric   

    actor_1_facebook_likes        gross  \
4                    131.0          NaN   
45                 26000.0  350034110.0   
89                   756.0  222487711.0   
12                   451.0  168368427.0   

                                               genres  ...  \
4                                         Documentary  ...   
45                              Action|Crime|Thriller  ...   
89  Action|Adventure|Animation|Comedy|Drama|Family...  ...   
12                                   Action|Adventure  ...   

   num_user_for_reviews language  country  content_rating       budget  \
4                   NaN      NaN      NaN             NaN          NaN   
45                657.0  English      USA           PG-13  190000000.0   
89                433.0  English      USA              PG  165000000.0   
12               1243.0  English       UK           PG-13  200000000.0   

    title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
4          NaN                   12.0        7.1           NaN   
45      2015.0                23000.0        7.2          2.35   
89      2014.0                  719.0        7.9          2.39   
12      2008.0                  412.0        6.7          2.35   

   movie_facebook_likes  
4                     0  
45                94000  
89                41000  
12                    0  

[4 rows x 28 columns]

Si vous n'avez pas compris cette partie, je vous recommande de relire la sélection de lignes et colonnes dans un tableau Numpy dans les notebooks 02_01 et 02_02.

# Séléctionner des lignes et des colonnes avec iloc

Nous souhaitons sélectionner les lignes 4 à 7, et les colonnes color et director_name. **iloc** permet la séléction avec les index ou plutôt position. NOus avons déjà ceux des lignes. Ceux des colonnes color et director_name sont 0 et 1. Nous pouvons donc faire notre sélection:

In [40]:
movie.iloc[4:8, 0:2]

color   director_name
4    NaN     Doug Walker
5  Color  Andrew Stanton
6  Color       Sam Raimi
7  Color    Nathan Greno

Reprenons le même exercice mais cette fois-ci avec les colonnes 3, 8, 6, 4:

In [41]:
movie.iloc[4:8, [3,8,6,4] ]

duration        gross     actor_2_name  director_facebook_likes
4       NaN          NaN       Rob Walker                    131.0
5     132.0   73058679.0  Samantha Morton                    475.0
6     156.0  336530303.0     James Franco                      0.0
7     100.0  200807262.0     Donna Murphy                     15.0

# Séléctionner des lignes et colonnes avec loc

Séléctionner avec les index c'est bien mais je préfère utiliser le nom des colonnes. C'est possible avec loc. loc permet de sélectionner avec le nom des lignes et des colonnes. Ici le nom de nos lignes est aussi l'index, on peut donc garder les nombres mais pour les colonnes nous utiliserons le nom:

Reproduisons ceci mais avec loc : 

In [42]:
movie.iloc[4:8, 0:2]

color   director_name
4    NaN     Doug Walker
5  Color  Andrew Stanton
6  Color       Sam Raimi
7  Color    Nathan Greno

On aura donc

In [43]:
movie.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [45]:
movie.loc[4:8, "color":"director_name"]

color   director_name
4    NaN     Doug Walker
5  Color  Andrew Stanton
6  Color       Sam Raimi
7  Color    Nathan Greno
8  Color     Joss Whedon

L'équivalent de ceci : 

In [46]:
movie.iloc[4:8, [3,8,6,4] ]

duration        gross     actor_2_name  director_facebook_likes
4       NaN          NaN       Rob Walker                    131.0
5     132.0   73058679.0  Samantha Morton                    475.0
6     156.0  336530303.0     James Franco                      0.0
7     100.0  200807262.0     Donna Murphy                     15.0

est : 

In [48]:
colonnes = movie.columns.tolist()
print(colonnes)

['color', 'director_name', 'num_critic_for_reviews', 'duration', 'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name', 'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name', 'movie_title', 'num_voted_users', 'cast_total_facebook_likes', 'actor_3_name', 'facenumber_in_poster', 'plot_keywords', 'movie_imdb_link', 'num_user_for_reviews', 'language', 'country', 'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes', 'imdb_score', 'aspect_ratio', 'movie_facebook_likes']


In [51]:
import numpy as np
np.array(colonnes)[[3,8,6,4]]

array(['duration', 'gross', 'actor_2_name', 'director_facebook_likes'],
      dtype='<U25')

In [52]:
movie.loc[4:8, ['duration', 'gross', 'actor_2_name', 'director_facebook_likes'] ]

duration        gross       actor_2_name  director_facebook_likes
4       NaN          NaN         Rob Walker                    131.0
5     132.0   73058679.0    Samantha Morton                    475.0
6     156.0  336530303.0       James Franco                      0.0
7     100.0  200807262.0       Donna Murphy                     15.0
8     141.0  458991599.0  Robert Downey Jr.                      0.0

# Filtrer les lignes d'un dataframe

Savoir filter les lignes d'un dataframe est une compétence importante quand on manipule des données tabulaires. Quelles sont les questions qui peuvent nous amener à filtrer les lignes d'un dataframe.

* Sélectionne les lignes (les films) dont le directeur est "Doug Walker" afin de plus tard calculer la note moyenne des films qu'il a dirigé

* Sélectionne les films francais dont la durée est inférieur à 1 heure 

* Sélectionne les films sortis avant 2005 ou les films dont le score est inférieur à 5.

Vous l'aurez compris, il faudra utiliser des conditions.

Les films dont le directeur est "Doug Walker" afin de plus tard calculer la note moyenne des films qu'il a dirigé

In [8]:
movie['director_name'] == "Doug Walker"

0       False
1       False
2       False
3       False
4        True
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13      False
14      False
15      False
16      False
17      False
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25      False
26      False
27      False
28      False
29      False
        ...  
4886    False
4887    False
4888    False
4889    False
4890    False
4891    False
4892    False
4893    False
4894    False
4895    False
4896    False
4897    False
4898    False
4899    False
4900    False
4901    False
4902    False
4903    False
4904    False
4905    False
4906    False
4907    False
4908    False
4909    False
4910    False
4911    False
4912    False
4913    False
4914    False
4915    False
Name: director_name, Length: 4916, dtype: bool

Cette condition nous retourne True quand la condition est vraie pour la ligne et False quand la condition est fausse.

Si nous utilisons cette condition pour indexer le dataframe, il nous retourne uniquement les lignes pour lesquelles on a True

In [9]:
condition = movie['director_name'] == "Doug Walker"
movie[condition]

color director_name  num_critic_for_reviews  duration  \
4   NaN   Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes actor_2_name  \
4                    131.0                     NaN   Rob Walker   

   actor_1_facebook_likes  gross       genres  ... num_user_for_reviews  \
4                   131.0    NaN  Documentary  ...                  NaN   

  language  country  content_rating budget  title_year actor_2_facebook_likes  \
4      NaN      NaN             NaN    NaN         NaN                   12.0   

  imdb_score  aspect_ratio movie_facebook_likes  
4        7.1           NaN                    0  

[1 rows x 28 columns]

Et voilà, on a apparemment un seul film dont Doug Walker est le directeur.

Parfois notre condition peut être composée de plusieurs autres conditions: 

Sélectionne les films francais dont la durée est inférieur à 1 heure. Ici, nous avons 2 conditions:
* langue == francais 
* duree < 1h

In [12]:
cond1 = movie['language'] == "French"
cond2 = movie['duration'] < 60

Vous serez tenté de faire ```python cond1 and cond2``` mais non. Avec Pandas **and** devient **&** et **or** devient **|**. 

In [15]:
cond1 and cond2

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [16]:
cond1 & cond2

0       False
1       False
2       False
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13      False
14      False
15      False
16      False
17      False
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25      False
26      False
27      False
28      False
29      False
        ...  
4886    False
4887    False
4888    False
4889    False
4890    False
4891    False
4892    False
4893    False
4894    False
4895    False
4896    False
4897    False
4898    False
4899    False
4900    False
4901    False
4902    False
4903    False
4904    False
4905    False
4906    False
4907    False
4908    False
4909    False
4910    False
4911    False
4912    False
4913    False
4914    False
4915    False
Length: 4916, dtype: bool

In [11]:
movie['duration']

0       178.0
1       169.0
2       148.0
3       164.0
4         NaN
5       132.0
6       156.0
7       100.0
8       141.0
9       153.0
10      183.0
11      169.0
12      106.0
13      151.0
14      150.0
15      143.0
16      150.0
17      173.0
18      136.0
19      106.0
20      164.0
21      153.0
22      156.0
23      186.0
24      113.0
25      201.0
26      194.0
27      147.0
28      131.0
29      124.0
        ...  
4886     79.0
4887     80.0
4888    100.0
4889     90.0
4890     90.0
4891    120.0
4892     91.0
4893    143.0
4894     85.0
4895     60.0
4896     88.0
4897     78.0
4898    108.0
4899    110.0
4900     90.0
4901     83.0
4902    111.0
4903     84.0
4904     82.0
4905     98.0
4906     77.0
4907     80.0
4908     81.0
4909     84.0
4910     95.0
4911     87.0
4912     43.0
4913     76.0
4914    100.0
4915     90.0
Name: duration, Length: 4916, dtype: float64

On peut maintenant indexer notre dataframe avec ces deux conditions : 

In [17]:
movie[cond1 & cond2]

color director_name  num_critic_for_reviews  duration  \
1123  Color           NaN                    23.0      52.0   
4791  Color  Chris Marker                    16.0      59.0   

      director_facebook_likes  actor_3_facebook_likes     actor_2_name  \
1123                      NaN                   114.0   Clotilde Hesme   
4791                    121.0                     0.0  Bertrand Cantat   

      actor_1_facebook_likes  gross                        genres  ...  \
1123                   164.0    NaN  Drama|Fantasy|Horror|Mystery  ...   
4791                    27.0    NaN                   Documentary  ...   

     num_user_for_reviews language  country  content_rating    budget  \
1123                 47.0   French   France           TV-MA       NaN   
4791                  4.0   French   France             NaN  150000.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
1123         NaN                  116.0        8.3         16.00   
4791      2004.0                    0.0        7.4          1.37   

     movie_facebook_likes  
1123                17000  
4791                   44  

[2 rows x 28 columns]

# Comment Créer un Dataframe ?

In [35]:
movie.describe()

num_critic_for_reviews     duration  director_facebook_likes  \
count             4867.000000  4901.000000              4814.000000   
mean               137.988905   107.090798               691.014541   
std                120.239379    25.286015              2832.954125   
min                  1.000000     7.000000                 0.000000   
25%                 49.000000    93.000000                 7.000000   
50%                108.000000   103.000000                48.000000   
75%                191.000000   118.000000               189.750000   
max                813.000000   511.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             4893.000000             4909.000000  4.054000e+03   
mean               631.276313             6494.488491  4.764451e+07   
std               1625.874802            15106.986884  6.737255e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                132.000000              607.000000  5.019656e+06   
50%                366.000000              982.000000  2.504396e+07   
75%                633.000000            11000.000000  6.110841e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     4.916000e+03                4916.000000           4903.000000   
mean      8.264492e+04                9579.815907              1.377320   
std       1.383222e+05               18164.316990              2.023826   
min       5.000000e+00                   0.000000              0.000000   
25%       8.361750e+03                1394.750000              0.000000   
50%       3.313250e+04                3049.000000              1.000000   
75%       9.377275e+04               13616.750000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           4895.000000  4.432000e+03  4810.000000   
mean             267.668846  3.654749e+07  2002.447609   
std              372.934839  1.002427e+08    12.453977   
min                1.000000  2.180000e+02  1916.000000   
25%               64.000000  6.000000e+06  1999.000000   
50%              153.000000  1.985000e+07  2005.000000   
75%              320.500000  4.300000e+07  2011.000000   
max             5060.000000  4.200000e+09  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             4903.000000  4916.000000   4590.000000           4916.000000  
mean              1621.923516     6.437429      2.222349           7348.294142  
std               4011.299523     1.127802      1.402940          19206.016458  
min                  0.000000     1.600000      1.180000              0.000000  
25%                277.000000     5.800000      1.850000              0.000000  
50%                593.000000     6.600000      2.350000            159.000000  
75%                912.000000     7.200000      2.350000           2000.000000  
max             137000.000000     9.500000     16.000000         349000.000000

In [36]:
movie['director_name'].describe()

count                 4814
unique                2397
top       Steven Spielberg
freq                    26
Name: director_name, dtype: object

Dans notre dataset, nous avons, 2397 directeurs uniques avec Steven Spielberg étant celui avec le plus de film qui est égale à 26

In [37]:
movie['director_name'].value_counts()

Steven Spielberg             26
Woody Allen                  22
Clint Eastwood               20
Martin Scorsese              20
Ridley Scott                 16
Spike Lee                    16
Renny Harlin                 15
Steven Soderbergh            15
Oliver Stone                 14
Tim Burton                   14
Barry Levinson               13
Robert Zemeckis              13
Robert Rodriguez             13
Ron Howard                   13
Joel Schumacher              13
Michael Bay                  12
Tony Scott                   12
Brian De Palma               12
Kevin Smith                  12
Chris Columbus               11
Rob Reiner                   11
Shawn Levy                   11
Sam Raimi                    11
Francis Ford Coppola         11
Richard Donner               11
Richard Linklater            11
John Carpenter               10
John McTiernan               10
Bobby Farrelly               10
David Fincher                10
                             ..
Chris At

In [38]:
movie.num_critic_for_reviews.describe()

count    4867.000000
mean      137.988905
std       120.239379
min         1.000000
25%        49.000000
50%       108.000000
75%       191.000000
max       813.000000
Name: num_critic_for_reviews, dtype: float64

In [39]:
movie.num_critic_for_reviews.std()

120.23937916791044

In [40]:
movie.mean(axis=0)

num_critic_for_reviews       1.379889e+02
duration                     1.070908e+02
director_facebook_likes      6.910145e+02
actor_3_facebook_likes       6.312763e+02
actor_1_facebook_likes       6.494488e+03
gross                        4.764451e+07
num_voted_users              8.264492e+04
cast_total_facebook_likes    9.579816e+03
facenumber_in_poster         1.377320e+00
num_user_for_reviews         2.676688e+02
budget                       3.654749e+07
title_year                   2.002448e+03
actor_2_facebook_likes       1.621924e+03
imdb_score                   6.437429e+00
aspect_ratio                 2.222349e+00
movie_facebook_likes         7.348294e+03
dtype: float64

In [41]:
movie.sum(axis=0)

num_critic_for_reviews       6.715920e+05
duration                     5.248520e+05
director_facebook_likes      3.326544e+06
actor_3_facebook_likes       3.088835e+06
actor_1_facebook_likes       3.188144e+07
gross                        1.931509e+11
num_voted_users              4.062825e+08
cast_total_facebook_likes    4.709438e+07
facenumber_in_poster         6.753000e+03
num_user_for_reviews         1.310239e+06
budget                       1.619785e+11
title_year                   9.631773e+06
actor_2_facebook_likes       7.952291e+06
imdb_score                   3.164640e+04
aspect_ratio                 1.020058e+04
movie_facebook_likes         3.612421e+07
dtype: float64

In [42]:
genre_un = movie.genres[0]
genre_un

'Action|Adventure|Fantasy|Sci-Fi'

In [43]:
genre_un.split('|')

['Action', 'Adventure', 'Fantasy', 'Sci-Fi']

In [44]:
movie['genres']

0                         Action|Adventure|Fantasy|Sci-Fi
1                                Action|Adventure|Fantasy
2                               Action|Adventure|Thriller
3                                         Action|Thriller
4                                             Documentary
5                                 Action|Adventure|Sci-Fi
6                                Action|Adventure|Romance
7       Adventure|Animation|Comedy|Family|Fantasy|Musi...
8                                 Action|Adventure|Sci-Fi
9                        Adventure|Family|Fantasy|Mystery
10                                Action|Adventure|Sci-Fi
11                                Action|Adventure|Sci-Fi
12                                       Action|Adventure
13                               Action|Adventure|Fantasy
14                               Action|Adventure|Western
15                        Action|Adventure|Fantasy|Sci-Fi
16                        Action|Adventure|Family|Fantasy
17            

In [45]:
movie.values

array([['Color', 'James Cameron', 723.0, ..., 7.9, 1.78, 33000],
       ['Color', 'Gore Verbinski', 302.0, ..., 7.1, 2.35, 0],
       ['Color', 'Sam Mendes', 602.0, ..., 6.8, 2.35, 85000],
       ...,
       ['Color', 'Benjamin Roberds', 13.0, ..., 6.3, nan, 16],
       ['Color', 'Daniel Hsia', 14.0, ..., 6.3, 2.35, 660],
       ['Color', 'Jon Gunn', 43.0, ..., 6.6, 1.85, 456]], dtype=object)

----

In [46]:
import numpy as np
import pandas as pd

In [47]:
movie.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

# comment creer un dataframe

In [48]:
noms = ['paul', 'radji', 'bob', 'alice']
ages = [41, 53, 56, 38]
poids = [12, 15.56, 98.25, 14.26]

In [49]:
data = {
    "noms": noms,
    "ages": ages,
    "poids": poids
}

In [50]:
data = pd.DataFrame(data)
data

noms  ages  poids
0   paul    41  12.00
1  radji    53  15.56
2    bob    56  98.25
3  alice    38  14.26

In [51]:
data.describe()

ages      poids
count   4.000000   4.000000
mean   47.000000  35.017500
std     8.831761  42.180653
min    38.000000  12.000000
25%    40.250000  13.695000
50%    47.000000  14.910000
75%    53.750000  36.232500
max    56.000000  98.250000

In [52]:
data.sum(axis=0)

noms     paulradjibobalice
ages                   188
poids               140.07
dtype: object

In [53]:
data = [
        {
            "nom": "paul",
            "age": 41,
         "poids": 12        },
        
        {
            "nom": "radji",
            "age": 53,
         "poids": 15.56        }
]

In [54]:
data

[{'nom': 'paul', 'age': 41, 'poids': 12},
 {'nom': 'radji', 'age': 53, 'poids': 15.56}]

In [55]:
pd.DataFrame(data)

age    nom  poids
0   41   paul  12.00
1   53  radji  15.56

In [56]:
movie.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

# Bonus: Importer les données d'un site web avec Pandas

In [53]:
!pip install lxml

     |████████████████████████████████| 6.3MB 2.1MB/s eta 0:00:01


In [3]:
tables = pd.read_html("https://fr.wikipedia.org/wiki/Liste_des_pr%C3%A9sidents_des_%C3%89tats-Unis_par_long%C3%A9vit%C3%A9")

In [4]:
tables[0]

0                                                  1  \
0  1.           Préciser le motif de la pose du bandeau.   
1  ou  Créer l'espace de discussion.(cette méthode es...   
2  2.               Informer les utilisateurs concernés.   

                                                   2  
0  Précisez le motif de la pose du bandeau en uti...  
1  Créez une section "Admissibilité" en page de d...  
2  Pensez à avertir le créateur de l'article, par...

In [5]:
tables[1]

No               Président               Né le Devientprésident le  \
0    1       George Washington  22 février 1732[1]       30 avril 1789   
1    2              John Adams  30 octobre 1735[1]         4 mars 1797   
2    3        Thomas Jefferson    13 avril 1743[1]         4 mars 1801   
3    4           James Madison     16 mars 1751[1]         4 mars 1809   
4    5            James Monroe       28 avril 1758         4 mars 1817   
5    6       John Quincy Adams     11 juillet 1767         4 mars 1825   
6    7          Andrew Jackson        15 mars 1767         4 mars 1829   
7    8        Martin Van Buren     5 décembre 1782         4 mars 1837   
8    9  William Henry Harrison      9 février 1773         4 mars 1841   
9   10              John Tyler        29 mars 1790        4 avril 1841   
10  11           James K. Polk     2 novembre 1795         4 mars 1845   
11  12          Zachary Taylor    24 novembre 1784         4 mars 1849   
12  13        Millard Fillmore      7 janvier 1800      9 juillet 1850   
13  14         Franklin Pierce    23 novembre 1804         4 mars 1853   
14  15          James Buchanan       23 avril 1791         4 mars 1857   
15  16         Abraham Lincoln     12 février 1809         4 mars 1861   
16  17          Andrew Johnson    29 décembre 1808       15 avril 1865   
17  18        Ulysses S. Grant       27 avril 1822         4 mars 1869   
18  19     Rutherford B. Hayes      4 octobre 1822         4 mars 1877   
19  20       James A. Garfield    19 novembre 1831         4 mars 1881   
20  21       Chester A. Arthur      5 octobre 1829   19 septembre 1881   
21  22        Grover Cleveland        18 mars 1837      4 mars 1885[4]   
22  23       Benjamin Harrison        20 août 1833         4 mars 1889   
23  24        Grover Cleveland        18 mars 1837      4 mars 1893[4]   
24  25        William McKinley     29 janvier 1843         4 mars 1897   
25  26      Theodore Roosevelt     27 octobre 1858   14 septembre 1901   
26  27     William Howard Taft   15 septembre 1857         4 mars 1909   
27  28          Woodrow Wilson    28 décembre 1856         4 mars 1913   
28  29       Warren G. Harding     2 novembre 1865         4 mars 1921   
29  30         Calvin Coolidge      4 juillet 1872         2 août 1923   
30  31          Herbert Hoover        10 août 1874         4 mars 1929   
31  32   Franklin D. Roosevelt     30 janvier 1882         4 mars 1933   
32  33         Harry S. Truman          8 mai 1884       12 avril 1945   
33  34    Dwight D. Eisenhower     14 octobre 1890     20 janvier 1953   
34  35         John F. Kennedy         29 mai 1917     20 janvier 1961   
35  36       Lyndon B. Johnson        27 août 1908    22 novembre 1963   
36  37           Richard Nixon      9 janvier 1913     20 janvier 1969   
37  38             Gerald Ford     14 juillet 1913         9 août 1974   
38  39            Jimmy Carter    1er octobre 1924     20 janvier 1977   
39  40           Ronald Reagan      6 février 1911     20 janvier 1981   
40  41       George H. W. Bush        12 juin 1924     20 janvier 1989   
41  42            Bill Clinton        19 août 1946     20 janvier 1993   
42  43          George W. Bush      6 juillet 1946     20 janvier 2001   
43  44            Barack Obama         4 août 1961     20 janvier 2009   
44  45            Donald Trump        14 juin 1946     20 janvier 2017   
45  46               Joe Biden    20 novembre 1942     20 janvier 2021   

           À l'âge de Cesse d'êtreprésident le À la retraitependant  \
0    57 ans, 67 jours              4 mars 1797                  NaN   
1   61 ans, 125 jours              4 mars 1801                  NaN   
2   57 ans, 325 jours              4 mars 1809                  NaN   
3   57 ans, 353 jours              4 mars 1817                  NaN   
4   58 ans, 310 jours              4 mars 1825                  NaN   
5   57 ans, 236 jours              4 mars 1829                  NaN   
6   61 ans, 354 jours       

https://fr.wikipedia.org/wiki/Liste_des_pr%C3%A9sidents_des_%C3%89tats-Unis_par_long%C3%A9vit%C3%A9